lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -1518,70 +1518,95 @@ namespace {
       return false;
     }
 
-    /// Check a reference to an entity within a global actor.
-    bool checkGlobalActorReference(
-        ConcreteDeclRef valueRef, SourceLoc loc, Type globalActor,
-        bool isCrossActor,
-        Expr *context) {
-      ValueDecl *value = valueRef.getDecl();
+    enum class AsyncMarkingResult {
+      FoundAsync, // successfully marked an implicitly-async operation
+      NotFound,  // fail: no valid implicitly-async operation was found
+      SyncContext, // fail: a valid implicitly-async op, but in sync context
+      NotConcurrentValue  // fail: valid op and context, but not ConcurrentValue
+    };
 
-      /// Returns true if this global-actor reference is acceptable because
-      /// it is part of an implicitly async operation, such as a call or
-      /// property access.
-      /// NOTE: This check will mutate the AST if it returns true!
-      auto inspectForImplicitlyAsync = [&] () -> bool {
-        // If our current context isn't an asynchronous one, don't
-        if (!isInAsynchronousContext())
-          return false;
+    /// Attempts to identify and mark a valid cross-actor use of a synchronous
+    /// actor-isolated member (e.g., sync function application, property access)
+    AsyncMarkingResult tryMarkImplicitlyAsync(SourceLoc declLoc,
```

Comment:
I like the factoring here, thanks!

Reply:
Oh this is awesome, I'll be able to add a case here for `FoundAsyncThrows` for a distributed func easily and cleanly then, rather than another ad-hoc thing ❤️

Reply:
Or maybe just Found... and applying both effects... either way, this is great